jax/api.py

```diff
@@ -1563,8 +1563,11 @@ def _vjp(fun: lu.WrappedFun, *primals, **kwargs):
     out_primal, out_vjp, aux = ad.vjp(flat_fun, primals_flat, has_aux=True)
     out_tree, aux_tree = out_aux_trees()
   out_primal_py = tree_unflatten(out_tree, out_primal)
-  vjp_py = partial(_vjp_pullback_wrapper, out_vjp,
-                   [_dtype(x) for x in out_primal], (out_tree, in_tree))
+  def vjp_py(py_args):
+      return Partial(partial(_vjp_pullback_wrapper,
+                             cotangent_dtypes=[_dtype(x) for x in out_primal],
+                             io_tree=(out_tree, in_tree)),
+                     fun=out_vjp)(py_args=py_args)
```

Comment:
I'm a little puzzled how this works. Doesn't the returned `vjp_py` function need to be a `Partial` object in order for it to be an output of the forward function?

Also: you're making a `Partial` object and then calling it immediately. That suggests you could just use a normal function call, because `Partial(f, x=...)(y=...)` is exactly equivalent to `f(x=..., y=...)`.

Reply:
Oh you're right.  It's broken, and the test I added fails.  I don't see how to use the existing `Partial` class to accomplish what we need to accomplish.  Do you want to go back to the version where I extend `Partial` to support static arguments, do you want me to reorder the parameters, or do you have another idea?

Reply:
Reordering arguments on the `_vjp_pullback_wrapper` helper so we can wrap it with `Partial`/`partial` seems like the cleanest solution to me, e.g., `_vjp_pullback_wrapper(cotangent_dtypes, io_tree, fun, py_args)` would put all the static arguments at the front, so this line could be just `vjp_py = Partial(partial(_vjp_pullback_wrapper, cotangent_dtypes, io_tree), out_vjp)`.

Reply:
Sounds good.  Done.